File: architecture-skills/design-patterns/examples/proxy/proxy.go
// Proxy pattern in Go (virtual proxy: lazy-loading image).
//
// An interface satisfied by both the real struct and the proxy; the
// proxy holds a nil pointer until first use. Go's implicit interface
// satisfaction means neither type declares that it implements Image —
// composition does all the work.

package main

import "fmt"

type Image interface {
	Display()
}

// RealImage is the real subject: expensive to create.
type RealImage struct {
	filename string
}

func NewRealImage(filename string) *RealImage {
	fmt.Printf("Loading %s from disk (expensive)\n", filename)
	return &RealImage{filename: filename}
}

func (r *RealImage) Display() {
	fmt.Printf("Displaying %s\n", r.filename)
}

// ImageProxy defers creating the RealImage until first use.
type ImageProxy struct {
	filename string
	real     *RealImage
}

func NewImageProxy(filename string) *ImageProxy {
	return &ImageProxy{filename: filename}
}

func (p *ImageProxy) Display() {
	if p.real == nil {
		p.real = NewRealImage(p.filename)
	}
	p.real.Display()
}

func main() {
	gallery := []Image{
		NewImageProxy("holiday.png"),
		NewImageProxy("unused.png"),
	}
	fmt.Println("Proxies created; nothing loaded yet")

	gallery[0].Display()
	gallery[0].Display() // loaded once, displayed twice

	fmt.Println("unused.png was never loaded")
}

File: architecture-skills/design-patterns/examples/proxy/proxy.rs
// Proxy pattern in Rust (virtual proxy: lazy-loading image).
//
// A trait object gives the shared subject interface, and OnceCell
// provides interior-mutable lazy initialization behind &self. Note that
// Rust's smart pointers are proxy-like by design: Deref lets Box, Rc,
// and friends transparently interpose on access to their target.

use std::cell::OnceCell;

trait Image {
    fn display(&self);
}

/// Real subject: expensive to create.
struct RealImage {
    filename: String,
}

impl RealImage {
    fn load(filename: &str) -> Self {
        println!("Loading {filename} from disk (expensive)");
        Self { filename: filename.to_string() }
    }
}

impl Image for RealImage {
    fn display(&self) {
        println!("Displaying {}", self.filename);
    }
}

/// Virtual proxy: defers creating RealImage until first use.
struct ImageProxy {
    filename: String,
    real: OnceCell<RealImage>,
}

impl ImageProxy {
    fn new(filename: &str) -> Self {
        Self { filename: filename.to_string(), real: OnceCell::new() }
    }
}

impl Image for ImageProxy {
    fn display(&self) {
        self.real.get_or_init(|| RealImage::load(&self.filename)).display();
    }
}

fn main() {
    let gallery: Vec<Box<dyn Image>> = vec![
        Box::new(ImageProxy::new("holiday.png")),
        Box::new(ImageProxy::new("unused.png")),
    ];
    println!("Proxies created; nothing loaded yet");

    gallery[0].display();
    gallery[0].display(); // loaded once, displayed twice

    println!("unused.png was never loaded");
}
